src/proto/batch/mod.rs

```diff
@@ -0,0 +1,338 @@
+
+mod cmd;
+
+pub use cmd::{CommitBatch, GetBatchStatus, OpenBatch};
+
+/// Batch of jobs.
+///
+/// Faktory guarantees a callback (`success` and/or `failure`) will be triggered after the execution
+/// of all the jobs belonging to the same batch has finished (successfully or with errors accordingly).
+/// The 'complete' callback will always be queued first.
+///
+/// Batches can be nested. They can also be re-opened, but - once a batch is committed - only those jobs
+/// that belong to this batch can re-open it.
+///
+/// An empty batch can be committed just fine. That will make Faktory immediately fire a callback (i.e. put
+/// the job specified in `complete` and/or the one specified in `success` onto the queues).
+///
+/// If you open a batch, but - for some reason - do not commit it within _30 minutes_, it will simply expire
+/// on the Faktory server (which means no callbackes will be fired).
+///
+/// Here is how you can create a simple batch:
+/// ```no_run
+/// # use faktory::Error;
+/// use faktory::{Producer, Job, Batch};
+///
+/// let mut prod = Producer::connect(None)?;
+/// let job1 = Job::builder("image").build();
+/// let job2 = Job::builder("image").build();
+/// let job_cb = Job::builder("clean_up").build();
+///
+/// let batch = Batch::builder("Image resizing workload".to_string()).with_complete_callback(job_cb);
+///
+/// let mut batch = prod.start_batch(batch)?;
+/// batch.add(job1)?;
+/// batch.add(job2)?;
+/// batch.commit()?;
+///
+/// # Ok::<(), Error>(())
+/// ```
+///
+/// Nested batches are also supported:
+/// ```no_run
+/// # use faktory::{Producer, Job, Batch, Error};
+/// # let mut prod = Producer::connect(None)?;
+/// let parent_job1 = Job::builder("stats_build").build();
+/// let parent_job2 = Job::builder("entries_update").build();
+/// let parent_cb = Job::builder("clean_up").build();
+///
+/// let child_job1 = Job::builder("image_recognition").build();
+/// let child_job2 = Job::builder("image_recognition").build();
+/// let child_cb = Job::builder("clean_up").build();
+///
+/// let parent_batch = Batch::builder("Image recognition and analysis workload".to_string()).with_complete_callback(parent_cb);
+/// let child_batch = Batch::builder("Image recognition workload".to_string()).with_success_callback(child_cb);
+///
+/// let mut parent = prod.start_batch(parent_batch)?;
+/// parent.add(parent_job1)?;
+/// parent.add(parent_job2)?;
+/// let mut child = parent.start_batch(child_batch)?;
+/// child.add(child_job1)?;
+/// child.add(child_job2)?;
+///
+/// child.commit()?;
+/// parent.commit()?;
+///
+/// # Ok::<(), Error>(())
+/// ```
+///
+/// In the example above, there is a single level nesting, but you can nest those batches as deep as you wish,
+/// effectively building a pipeline this way, since the Faktory guarantees that callback jobs will not be queued unless
+/// the batch gets committed.
+///
+/// You can retieve the batch status using a [`tracker`](struct.Tracker.html):
```

Comment:
```suggestion
/// You can retieve the batch status using a [`Tracker`]:
```

Reply:
Uppercased, but looks like `struct.Tracker.html` link will still be needed since `Tracker` is not in the scope.

Reply:
Ah, a nice trick here is:
```rust
#[cfg(doc)]
use whatever::Tracker;
```
to bring it in scope so rustdoc can find it.

Reply:
This is clever. Updated